pulp_container/app/registry_api.py

```diff
@@ -803,93 +793,157 @@ def put(self, request, path, pk=None):
         if request.content_type not in (
             models.MEDIA_TYPE.MANIFEST_V2,
             models.MEDIA_TYPE.MANIFEST_OCI,
+            models.MEDIA_TYPE.MANIFEST_LIST,
+            models.MEDIA_TYPE.INDEX_OCI,
         ):
-            # we suport only v2 docker/oci schema upload
             raise ManifestInvalid(digest=manifest_digest)
-        # both docker/oci format should contain config, digest, mediaType, size
-        config_layer = content_data.get("config")
-        try:
-            config_digest = config_layer.get("digest")
-            config_blob = models.Blob.objects.get(digest=config_digest)
-        except models.Blob.DoesNotExist:
-            raise BlobNotFound(digest=config_digest)
-        config_media_type = config_layer.get("mediaType")
-        if config_media_type not in (
-            models.MEDIA_TYPE.CONFIG_BLOB,
-            models.MEDIA_TYPE.CONFIG_BLOB_OCI,
+
+        if request.content_type in (
+            models.MEDIA_TYPE.MANIFEST_LIST,
+            models.MEDIA_TYPE.INDEX_OCI,
         ):
-            raise BlobInvalid(digest=config_blob.digest)
+            manifest_list = models.Manifest(
+                digest=manifest_digest,
+                schema_version=content_data["schemaVersion"],
+                media_type=content_data.get("mediaType", models.MEDIA_TYPE.INDEX_OCI),
+            )
+            try:
+                manifest_list.save()
+            except IntegrityError:
+                manifest_list = models.Manifest.objects.get(digest=manifest_list.digest)
+                manifest_list.touch()
 
-        manifest = models.Manifest(
-            digest=manifest_digest,
-            schema_version=2,
-            media_type=request.content_type,
-            config_blob=config_blob,
-        )
-        try:
-            manifest.save()
-        except IntegrityError:
-            manifest = models.Manifest.objects.get(digest=manifest.digest)
-            manifest.touch()
-        ca = ContentArtifact(artifact=artifact, content=manifest, relative_path=manifest.digest)
-        try:
-            ca.save()
-        except IntegrityError:
-            ca = ContentArtifact.objects.get(content=manifest, relative_path=manifest.digest)
+            ca = ContentArtifact(
+                artifact=artifact, content=manifest_list, relative_path=manifest_list.digest
+            )
+            try:
+                ca.save()
+            except IntegrityError:
+                ca = ContentArtifact.objects.get(
+                    content=manifest_list, relative_path=manifest_list.digest
+                )
+                if not ca.artifact:
+                    ca.artifact = artifact
+                    ca.save(update_fields=["artifact"])
+
+            manifests_to_list = []
+            for manifest_metadata in content_data.get("manifests"):
+                manifest = models.Manifest.objects.get(digest=manifest_metadata["digest"])
```

Comment:
If we want to let users know what manifests are missing in the database, then I am all for it (+10 lines of code).

Reply:
I was thinking something along the lines:
```
mlm = {}
for manifest_metadata in content_data.get("manifests"):
    mlm[manifest_metadata["digest"]] = manifest_metadata["platform"]
manifests_qs = models.Manifest.objects.filter(digest=mlm.keys())
manifests_to_list = []
for m in manifests_qs:
    manifest_to_list = models.ManifestListManifest(
        manifest_list=m,
        image_manifest=manifest_list,
        architecture=mlm[m.difest]["architecture"],
        os=mlm[m.digest]["os"],
        features=mlm[m.digest].get("features", ""),
        variant=mlm[m.digest].get("variant", ""),
        os_version=mlm[m.digest].get("os.version", ""),
        os_features=mlm[m.digest].get("os.features", ""),
    )
        manifests_to_list.append(manifest_to_list)

models.ManifestListManifest.objects.bulk_create(
    manifests_to_list, ignore_conflicts=True, batch_size=1000
)
```

Reply:
:thinking:. Also, in addition to that, I will need to add a set difference operation to find out which manifests were not found in the database and raise an error.

Reply:
yeah..so here is the thing that i do want to avoid this error raising. if you look at blobs from image manifest we are doing same thing.
I am not 100% positive on whether this is a correct approach. One would need to open docker/distribution code and find whether registry does any checks when it PUTs the manifest json.